### PR TITLE
Add option for Node.js require resolution algorithm

### PR DESCRIPTION
**Pull request by kirbysayshi**
Reopened after returning! Redux of #42.

Hope this is adequate! I'll let you do the version bump etc.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -22,5 +27,15 @@ module.exports = function (grunt, opts) {
 		return result.concat(Array.isArray(deps) ? deps : Object.keys(deps));
 	}, []);
 
-	multimatch(names, pattern).forEach(grunt.loadNpmTasks);
+	var resolver = resolveFrom.bind(null, process.cwd());
```
Comment (sindresorhus) [inline index.js]: Use https://github.com/sindresorhus/resolve-cwd

File: index.js [hunk 2]
```diff
@@ -22,5 +27,15 @@ module.exports = function (grunt, opts) {
...
+
+	multimatch(names, pattern).forEach(function (pkgName) {
+		if (config.nodeModulesResolution) {
+			var resolved = resolver(path.join(pkgName, 'package.json'));
+			var tasksDir = path.join(path.dirname(resolved), 'tasks');
+			grunt.loadTasks(tasksDir);
```
Comment (sindresorhus) [inline index.js]: Not sure why you're over-complicating it. Can just be:

```js
grunt.loadTasks(resolveCwd(path.join(pkgName, 'tasks')));
```

File: readme.md [hunk 3]
```diff
@@ -102,35 +102,46 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
+
 ### All options in use
 
 ```js
 require('load-grunt-tasks')(grunt, {
 	pattern: 'grunt-contrib-*',
 	config: '../package.json',
-	scope: 'devDependencies'
+	scope: 'devDependencies',
+  nodeModulesResolution: true
```
Comment (sindresorhus) [inline readme.md]: wrong indent

File: readme.md [hunk 4]
```diff
@@ -102,35 +102,46 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
 });
 ```
 
-
 ## Options
 
 ### pattern
 
-Type: `string`, `array`
```
Comment (sindresorhus) [inline readme.md]: Don't remove

File: readme.md [hunk 5]
```diff
@@ -102,35 +102,46 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
 require('load-grunt-tasks')(grunt, {
 	pattern: 'grunt-contrib-*',
 	config: '../package.json',
-	scope: 'devDependencies'
+	scope: 'devDependencies',
+  nodeModulesResolution: true
 });
 ```
 
-
```
Comment (sindresorhus) [inline readme.md]: Don't remove

File: readme.md [hunk 6]
```diff
@@ -102,35 +102,46 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
 
 ### pattern
 
-Type: `string`, `array`  
+Type: `string`, `array`
 Default: `['grunt-*', '@*/grunt-*']` ([globbing pattern](https://github.com/isaacs/minimatch))
 
 ### config
 
-Type: `string`, `object`
```
Comment (sindresorhus) [inline readme.md]: Don't remove

File: readme.md [hunk 7]
```diff
@@ -102,35 +102,46 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
 
 ### config
 
-Type: `string`, `object`  
+Type: `string`, `object`
 Default: Path to nearest package.json
 
 ### scope
 
-Type: `string`, `array`
```
Comment (sindresorhus) [inline readme.md]: Don't remove

File: readme.md [hunk 8]
```diff
@@ -102,35 +102,46 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
-Default: `['dependencies', 'devDependencies', 'peerDependencies', 'optionalDependencies']`  
+Type: `string`, `array`
+Default: `['dependencies', 'devDependencies', 'peerDependencies', 'optionalDependencies']`
 Values: `'dependencies'`, `'devDependencies'`, `'peerDependencies'`, `'optionalDependencies'`, `'bundledDependencies'`
 
+### nodeModulesResolution
+
+Type: `boolean`
+Default: `true`
+Values: `true`, `false`
```
Comment (sindresorhus) [inline readme.md]: No need to use `Values` here.

File: readme.md [hunk 9]
```diff
@@ -102,35 +102,46 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
-Type: `string`, `array`  
-Default: `['dependencies', 'devDependencies', 'peerDependencies', 'optionalDependencies']`  
+Type: `string`, `array`
+Default: `['dependencies', 'devDependencies', 'peerDependencies', 'optionalDependencies']`
 Values: `'dependencies'`, `'devDependencies'`, `'peerDependencies'`, `'optionalDependencies'`, `'bundledDependencies'`
 
+### nodeModulesResolution
+
+Type: `boolean`
+Default: `true`
```
Comment (sindresorhus) [inline readme.md]: It should be opt-in as it's not how grunt works by default.

File: readme.md [hunk 10]
```diff
@@ -102,35 +102,46 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
 
 ### scope
 
-Type: `string`, `array`  
-Default: `['dependencies', 'devDependencies', 'peerDependencies', 'optionalDependencies']`  
+Type: `string`, `array`
+Default: `['dependencies', 'devDependencies', 'peerDependencies', 'optionalDependencies']`
 Values: `'dependencies'`, `'devDependencies'`, `'peerDependencies'`, `'optionalDependencies'`, `'bundledDependencies'`
 
+### nodeModulesResolution
```
Comment (sindresorhus) [inline readme.md]: I think `requireResolution` would be a better name.

File: index.js (hunk 2 above)
Comment (sindresorhus) [inline index.js]: Just override the `pkgName` variable in the if statement instead of having two calls to grunt.loadTasks.

Comment (kirbysayshi) [inline index.js]: I was under the impression that resolveCwd returned the main file of a package, not the folder.

Comment (kirbysayshi) [inline index.js]: Also I prefer to not reassign variables, especially those passed in as arguments. But I'll change the if statement.

Comment (sindresorhus) [inline index.js]: Oh, never mind, forgot grunt tasks doesn't have a `main`...

**Comment by kirbysayshi** [comment]
Turns out two things are broken.

One is that I was reading from `config` and not `opts`.

the other is that resolve-from seems to be completely broken:

```
$ ls -la node_modules | grep grunt
drwxr-xr-x   11 drewp  staff    374 Dec  9 15:59 grunt
drwxr-xr-x   12 drewp  staff    408 Dec  9 15:59 grunt-cli
drwxr-xr-x   13 drewp  staff    442 Dec  9 15:59 grunt-legacy-log
drwxr-xr-x   12 drewp  staff    408 Dec  9 15:59 grunt-legacy-log-utils
drwxr-xr-x   11 drewp  staff    374 Dec  9 15:59 grunt-legacy-util
drwxr-xr-x    6 drewp  staff    204 Dec  9 15:59 grunt-svgmin
$ node --version
v5.1.1
$ node
> var rf = require('resolve-from')
undefined
> rf('.', 'grunt-svgmin')
null
```

So I'm going to switch back to require-resolve.

**Comment by sindresorhus** [comment]
No, it is correct. grunt-svgmin doesn't have a main entry point. Same with the core resolve:

```
~/dev/load-grunt-tasks master
❯ node              
> require.resolve('grunt-svgmin')
Error: Cannot find module 'grunt-svgmin'
```

**Comment by kirbysayshi** [comment]
yes, I just noticed this too. newest commit has a workaround.

**Comment by kirbysayshi** [comment]
ok, should be good to go now.

File: readme.md [hunk 11]
```diff
@@ -4,7 +4,7 @@
 
 Usually you would have to load each task one by one, which is unnecessarily cumbersome.
 
-This module will read the `dependencies`/`devDependencies`/`peerDependencies`/`optionalDependencies` in your package.json and load grunt tasks that match the provided patterns.
+This module will read the `dependencies`/`devDependencies`/`peerDependencies`/`optionalDependencies` in your package.json and load grunt tasks that match the provided patterns. It will also traverse the file hierarchy, just like node does when requiring modules (can be disabled).
```
Comment (sindresorhus) [inline readme.md]: This added sentence can be moved to the option on line 144 and reworded.

File: readme.md [hunk 12]
```diff
@@ -102,13 +102,20 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
...
+
 ### All options in use
 
 ```js
 require('load-grunt-tasks')(grunt, {
 	pattern: 'grunt-contrib-*',
 	config: '../package.json',
-	scope: 'devDependencies'
+	scope: 'devDependencies',
+	requireResolution: true
```
Comment (sindresorhus) [inline readme.md]: This needs to be updated.

File: readme.md [hunk 13]
```diff
@@ -102,13 +102,20 @@ require('load-grunt-tasks')(grunt, {scope: 'devDependencies'});
 require('load-grunt-tasks')(grunt, {scope: ['devDependencies', 'dependencies']});
 ```
 
+### Only load from `node_modules` in the current working directory
```
Comment (sindresorhus) [inline readme.md]: This needs to be made inverse.

File: index.js [hunk 14]
```diff
@@ -22,5 +27,16 @@ module.exports = function (grunt, opts) {
...
+		if (opts.requireResolution) {
+			// This resolution is complicated because most grunt plugins are written
+			// in violation of package.json conventions. And example is not having a
+			// `main` field defined, which will cause `require` or `resolve`
+			// to fail. So better to lookup a guaranteed file, such as package.json.
+			var pkg = resolveCwd(path.join(pkgName, 'package.json'));
+			var root = path.dirname(pkg);
+			pkgName = path.join(root, 'tasks');
+		}
+		grunt.loadTasks(pkgName);
```
Comment (sindresorhus) [inline index.js]: Add an empty line above here.

File: index.js [hunk 15]
```diff
@@ -22,5 +27,16 @@ module.exports = function (grunt, opts) {
 		return result.concat(Array.isArray(deps) ? deps : Object.keys(deps));
 	}, []);
 
-	multimatch(names, pattern).forEach(grunt.loadNpmTasks);
+	multimatch(names, pattern).forEach(function (pkgName) {
+		if (opts.requireResolution) {
```
Comment (sindresorhus) [inline index.js]: `if (opts.requireResolution === true) {` here and you can drop line 19-21.

File: readme.md [hunk 16]
```diff
@@ -131,6 +138,10 @@ Type: `string`, `array`
 Default: `['dependencies', 'devDependencies', 'peerDependencies', 'optionalDependencies']`  
 Values: `'dependencies'`, `'devDependencies'`, `'peerDependencies'`, `'optionalDependencies'`, `'bundledDependencies'`
 
+### requireResolution
+
+Type: `boolean`
+Default: `false`
```
Comment (sindresorhus) [inline readme.md]: Option needs a description.

**Comment by kirbysayshi** [comment]
ok, should be good!

**Comment by sindresorhus** [comment]
Thanks! Just published a new release ;)

**Pull request closed without merging**